### PR TITLE
core/src/transport/memory: Add basic dialer listener test

### PR DESCRIPTION
**Pull request by mxinden**
Adds a basic test using two memory transport instances, one dialing the
other, sending a single message over the /wire/.

### CODE REVIEW

**Comment by mxinden** [comment]
@romanb @twittner @tomaka friendly ping. What do you think of this pull request?

File: core/src/transport/memory.rs [hunk 1]
```diff
@@ -308,5 +311,45 @@ mod tests {
...
     }
 
-    // TODO: test that is actually works
+    #[test]
+    fn communicating_between_dialer_and_listener() {
+        let msg = [1, 2, 3];
+
+        // Setup listener.
+
+        let rand_port: NonZeroU64;
```
Comment (romanb) [inline core/src/transport/memory.rs]:
```suggestion
        let port = std::ord::max(1, rand::random::<u64>());
```

File: core/src/transport/memory.rs [hunk 2]
```diff
@@ -308,5 +310,37 @@ mod tests {
...
     }
 
-    // TODO: test that is actually works
+    #[test]
+    fn communicating_between_dialer_and_listener() {
+        let msg = [1, 2, 3];
+
+        // Setup listener.
+
+        let t1_addr: Multiaddr = "/memory/1".parse().unwrap();
```
[the remark this answers is not in the document]
Comment (romanb) [reply core/src/transport/memory.rs]: > Good point. I decided for randomizing the port to keep the test simple. I think we can take the risk of not hitting a collision of two _random_ `u64`'s, right?

_Probably_ :smile:

**Comment by mxinden** [comment]
Delaying this pull request until the *stable-future* branch was merged into *master*.

**Comment by tomaka** [comment]
Let's merge this now.

**Comment by tomaka** [comment]
Compilation errors. Branch needs an update.

File: core/src/transport/memory.rs [hunk 3]
```diff
@@ -311,5 +313,43 @@ mod tests {
...
     }
 
-    // TODO: test that is actually works
+    #[test]
+    fn communicating_between_dialer_and_listener() {
+        let msg = [1, 2, 3];
+
+        // Setup listener.
+
+        let rand_port = rand::random::<u64>() + 1;
```
Comment (tomaka) [inline core/src/transport/memory.rs]: Why this `+ 1` by the way?

Comment (mxinden) [reply core/src/transport/memory.rs]: Well, ... `0` would instruct the OS to bind to any port. I guess it is very unlikely. Do you want me to remove it @tomaka?

Comment (tomaka) [reply core/src/transport/memory.rs]: Well, it is equally likely that we randomly generate `u64::max_value()` and that the `+1` either panics or wraps to 0.

Comment (romanb) [reply core/src/transport/memory.rs]: I would go with `saturating_add(1)`.

Comment (tomaka) [reply core/src/transport/memory.rs]: Here's the "Apply suggestion" button ;)

```suggestion
        let rand_port = rand::random::<u64>().saturating_add(1);
```

Comment (mxinden) [reply core/src/transport/memory.rs]: Thanks @romanb and @tomaka!

**Pull request merged**